### PR TITLE
Convert local scheduler messages to flatbuffers

### PR DESCRIPTION
**Pull request by pcmoritz**

### CODE REVIEW

File: src/plasma/plasma_extension.cc [hunk 1]
```diff
@@ -357,26 +355,30 @@ PyObject *PyPlasma_receive_notification(PyObject *self, PyObject *args) {
    * object was added, return a tuple of its fields: ObjectID, data_size,
    * metadata_size. If the object was deleted, data_size and metadata_size will
    * be set to -1. */
-  int nbytes =
-      read_bytes(plasma_sock, (uint8_t *) &object_info, sizeof(object_info));
+  int64_t size;
+  int nbytes = read_bytes(plasma_sock, (uint8_t *) &size, sizeof(size));
+  CHECK(nbytes == sizeof(size));
```
Comment (robertnishihara) [inline src/plasma/plasma_extension.cc]: All of these `nbytes` should not be `nbytes` bcause `read_bytes returns an error code, so this check should fail.

File: src/plasma/plasma_manager.cc [hunk 2]
```diff
@@ -1346,11 +1348,13 @@ void process_object_notification(event_loop *loop,
                                  void *context,
                                  int events) {
   PlasmaManagerState *state = (PlasmaManagerState *) context;
-  ObjectInfo object_info;
   /* Read the notification from Plasma. */
-  int error =
-      read_bytes(client_sock, (uint8_t *) &object_info, sizeof(object_info));
-  if (error < 0) {
+  int64_t size;
+  int nbytes = read_bytes(client_sock, (uint8_t *) &size, sizeof(int64_t));
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: These `nbytes` should not be `nbytes`.

File: src/local_scheduler/local_scheduler.cc [hunk 3]
```diff
@@ -622,77 +637,64 @@ void process_message(event_loop *loop,
...
-  case REGISTER_WORKER_INFO: {
+  case MessageType_RegisterWorkerInfo: {
     /* Update the actor mapping with the actor ID of the worker (if an actor is
      * running on the worker). */
-    register_worker_info *info =
-        (register_worker_info *) utarray_front(state->input_buffer);
-    if (!ActorID_equal(info->actor_id, NIL_ACTOR_ID)) {
+    auto message = flatbuffers::GetRoot<RegisterWorkerInfo>(
+        utarray_front(state->input_buffer));
+    int64_t worker_pid = message->worker_pid();
```
Comment (robertnishihara) [inline src/local_scheduler/local_scheduler.cc]: you use `int64_t` here, but in the flatbuffer spec it is `ulong`.

We should make them consistent.

File: src/plasma/format/plasma.fbs [hunk 4]
```diff
@@ -39,7 +39,9 @@ enum MessageType:int {
   PlasmaUnsubscribeRequest,
   // Sending and receiving data.
```
Comment (robertnishihara) [inline src/plasma/format/plasma.fbs]: It'd be good to expand on the documentation for `PlasmaDataRequest` and `PlasmaDataReply`. For example, describe how this is used for transferring objects between plasma managers, and how there will be one `PlasmaDataRequest` to initiate the transfer and then multiple `PlasmaDataReply` messages containing the bytes.

File: src/plasma/plasma.cc [hunk 5]
```diff
@@ -20,3 +20,13 @@ void warn_if_sigpipe(int status, int client_sock) {
   }
   LOG_FATAL("Failed to write message to client on fd %d.", client_sock);
 }
+
+uint8_t *create_object_info_buffer(ObjectInfoT *object_info) {
```
Comment (robertnishihara) [inline src/plasma/plasma.cc]: Please document this saying that it mallocs and that it's the callers responsibility to free the returned buffer.

File: src/plasma/plasma_manager.cc [hunk 6]
```diff
@@ -1311,33 +1312,35 @@ void log_object_hash_mismatch_error_object_callback(ObjectID object_id,
 }
 
 void process_add_object_notification(PlasmaManagerState *state,
-                                     ObjectInfo object_info) {
-  ObjectID obj_id = object_info.obj_id;
+                                     ObjectID object_id,
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: Why do you pass in the arguments here individually instead of passing in the `ObjectInfo` object? Just curious.

File: src/plasma/plasma_manager.cc [hunk 7]
```diff
@@ -1346,10 +1349,21 @@ void process_object_notification(event_loop *loop,
                                  void *context,
                                  int events) {
   PlasmaManagerState *state = (PlasmaManagerState *) context;
-  ObjectInfo object_info;
   /* Read the notification from Plasma. */
-  int error =
-      read_bytes(client_sock, (uint8_t *) &object_info, sizeof(object_info));
+  int64_t size;
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: This construct where we call `read_bytes` twice (once to get the size and once to get the data) feels a bit too low-level to be here. It also appears in both the plasma manager and in the local scheduler, which makes me think it should either be in the plasma client library or in `io.cc`. What do you think?

File: src/plasma/plasma_manager.cc (hunk 6 above)
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: Oh is it to avoid copying the `ObjectInfo` buffer?

File: src/plasma/plasma_store.cc [hunk 8]
```diff
@@ -145,8 +148,30 @@ PlasmaStoreState *PlasmaStoreState_init(event_loop *loop,
   return state;
 }
 
+void PlasmaStoreState_free(PlasmaStoreState *state) {
```
Comment (robertnishihara) [inline src/plasma/plasma_store.cc]: Why is this necessary now whereas it wasn't before? What changed?

File: src/plasma/plasma_store.cc [hunk 9]
```diff
@@ -213,10 +238,9 @@ int create_object(Client *client_context,
   get_malloc_mapinfo(pointer, &fd, &map_size, &offset);
   assert(fd != -1);
 
-  entry = (object_table_entry *) malloc(sizeof(object_table_entry));
-  memset(entry, 0, sizeof(object_table_entry));
-  memcpy(&entry->object_id, &obj_id, sizeof(entry->object_id));
-  entry->info.obj_id = obj_id;
+  entry = new object_table_entry();
```
Comment (robertnishihara) [inline src/plasma/plasma_store.cc]: Ok, we can call `new` because we're sticking entry in a hash table which will never try to move or copy or delete the entry. If we were sticking it in a `utarray` then we would have problems when the array gets resized, right?

File: src/plasma/plasma_store.cc [hunk 10]
```diff
@@ -598,12 +620,14 @@ void remove_objects(PlasmaStoreState *plasma_state,
...
-                       ObjectInfo *notification) {
+                       ObjectInfoT *object_info) {
   NotificationQueue *queue, *temp_queue;
   HASH_ITER(hh, plasma_state->pending_notifications, queue, temp_queue) {
-    utarray_push_back(queue->object_notifications, notification);
+    uint8_t *notification = create_object_info_buffer(object_info);
+    utarray_push_back(queue->object_notifications, &notification);
     send_notifications(plasma_state->loop, queue->subscriber_fd, plasma_state,
                        0);
+    // free(notification);
```
Comment (robertnishihara) [inline src/plasma/plasma_store.cc]: Can you document when this gets freed? And get rid of the comment.

File: src/plasma/plasma_store.cc [hunk 11]
```diff
@@ -650,6 +676,7 @@ void send_notifications(event_loop *loop,
       }
     }
     num_processed += 1;
+    free(data);
```
Comment (robertnishihara) [inline src/plasma/plasma_store.cc]: Can you document where the corresponding `malloc` happened? I believe it is in `create_object_info_buffer` within `push_notification`.

File: src/plasma/plasma_store.cc [hunk 12]
```diff
@@ -622,14 +646,16 @@ void send_notifications(event_loop *loop,
...
+    uint8_t **notification =
+        (uint8_t **) utarray_eltptr(queue->object_notifications, i);
+    uint8_t *data = *notification;
+    /* Decode the length, which is the first bytes of the message. */
+    int64_t size = *((int64_t *) data);
 
     /* Attempt to send a notification about this object ID. */
-    int nbytes = send(client_sock, (char const *) notification,
-                      sizeof(*notification), 0);
+    int nbytes = send(client_sock, data, size + sizeof(int64_t), 0);
```
Comment (robertnishihara) [inline src/plasma/plasma_store.cc]: I have a slight preference for `sizeof(int64_t) + size` since the size is getting sent first.

File: src/plasma/plasma_store.cc [hunk 13]
```diff
@@ -622,14 +646,16 @@ void send_notifications(event_loop *loop,
...
     if (nbytes >= 0) {
-      CHECK(nbytes == sizeof(*notification));
+      CHECK(nbytes == size + sizeof(int64_t));
```
Comment (robertnishihara) [inline src/plasma/plasma_store.cc]: same down here.

File: src/plasma/test/manager_tests.cc [hunk 14]
```diff
@@ -258,31 +258,38 @@ TEST object_notifications_test(void) {
   CHECK(fcntl(fd[1], F_SETFL, flags | O_NONBLOCK) == 0);
 
   ObjectID oid = globally_unique_id();
```
Comment (robertnishihara) [inline src/plasma/test/manager_tests.cc]: I realize you didn't introduce it in this PR, but can you change `oid` to `object_id`?

**Review by robertnishihara — COMMENTED** [review]
This looks very nice!

File: src/common/CMakeLists.txt [hunk 15]
```diff
@@ -19,7 +19,7 @@ add_custom_target(gen_common_fbs ALL)
 
 add_custom_command(
   TARGET gen_common_fbs
-  COMMAND ${FLATBUFFERS_COMPILER} -c -o ${OUTPUT_DIR} ${COMMON_FBS_SRC}
+  COMMAND ${FLATBUFFERS_COMPILER} -c -o ${OUTPUT_DIR} ${COMMON_FBS_SRC} --gen-object-api
```
Comment (robertnishihara) [inline src/common/CMakeLists.txt]: What does `--gen-object-api` do? Maybe add a comment explaining?

File: src/common/format/common.fbs [hunk 16]
```diff
@@ -49,4 +49,22 @@ table TaskInfo {
...
+  // Number of bytes the metadata of this object occupies in memory.
+  metadata_size: long;
+  // Unix epoch of when this object was created.
+  create_time: long;
+  // How long creation of this object took.
+  construct_duration: long;
+  // Hash of the object content.
+  digest: string;
+  // Specifies if this object was deleted or added.
+  is_deletion: bool;
```
Comment (pcmoritz) [inline src/common/format/common.fbs]: Yeah, that's right. We had it as part of the original object_info and it is easier to not have another message that's why I kept it.

File: src/plasma/plasma_manager.cc (hunk 6 above)
Comment (pcmoritz) [inline src/plasma/plasma_manager.cc]: It is to make the code independent of the concrete ObjectInfoT representation (the main point being that ObjectInfoT uses std::string to represent object_ids and this method should not be aware of that fact).

File: src/plasma/plasma_manager.cc (hunk 7 above)
Comment (pcmoritz) [inline src/plasma/plasma_manager.cc]: Yes, I was thinking about this, the only reason why I didn't do it is that the error handling is slightly different in 2 of these 3 cases. I'll change it and decouple the error handling by returning a status code.

File: src/plasma/plasma_store.cc (hunk 8 above)
Comment (pcmoritz) [inline src/plasma/plasma_store.cc]: valgrind was complaining somehow and that fixed it, i didn't investigate it.

File: src/plasma/plasma_store.cc (hunk 9 above)
Comment (pcmoritz) [inline src/plasma/plasma_store.cc]: Yes, we should really switch to STL containers sooner rather than later. This caused lots of headaches for me and will cause more for other people I feel. The problem here is that if we use malloc, C++ constructors don't get called, which is very bad if there are std::strings anywhere, as is the case for ObjectInfoT.
